Add distance weighted particle interpolator

gassmoeller opened the pull request:
This is not ready for review.

@anne-glerum: This PR adds a new particle interpolator that interpolates between particles based on a distance weighted average (similar to what radial basis functions would do). The interpolator in its current form produces worse errors (though the same convergence rate) as the 'cell average' interpolator. However, it should be smoother and less affected by small particle movements across cell boundaries.

 This PR is currently in a very rough state, I only open this to remind myself to finish it after the hackathon.

gassmoeller commented:
I cleaned up the PR and made sure I can run the particle benchmarks to show the method is converging. In terms of accuracy this interpolator has a similar convergence rate as 'cell average', but because the particle contribution is weighted by distance and particles of neighbor cells are considered as well, it should be less sensitive to particles crossing cell boundaries.

Ready for review.

MFraters commented on include/aspect/particle/interpolator/distance_weighted_average.h:
```diff
@@ -0,0 +1,72 @@
+    {
+      /**
+       * Return the interpolated properties of all particles of the given cell using bilinear least squares method.
+       * Currently, only the two dimensional model is supported.
```
You say it work in 3d, but here it is stated it only works in 2d. Can you update this?

MFraters commented on include/aspect/particle/interpolator/distance_weighted_average.h:
```diff
@@ -0,0 +1,72 @@
+                               const typename parallel::distributed::Triangulation<dim>::active_cell_iterator &cell) const override;
+
+          // avoid -Woverloaded-virtual:
+          using Interface<dim>::properties_at_points;
```
Can you check if you still need this?

MFraters reviewed (approved):
Some small comments, otherwise looks good to me.

tjhei reviewed (approved):
Looks good. I have a few questions to make sure I understand things correctly.

gassmoeller commented:
I think I have addressed all comments, but let me run some more benchmarks with this before merging.

gassmoeller commented:
Ok, I am done benchmarking. I think this is ready. In the benchmark `rigid_shear/transient` it reaches the same convergence order as `cell average` at better accuracy:
![error_over_resolution_end](https://github.com/geodynamics/aspect/assets/7582930/ae50c44d-df88-4bd5-9a68-57e5cae1f85b)

If anyone used `cell average` regularly before, I would now recommend using `distance weighted average` instead.

naliboff commented:
@gassmoeller - thanks for the work to add this feature! 

I did a bit of testing related to #4370 and ran into an issue when doing an initial adaptive refinement step on the first time step. 

The model runs without issue if `set Initial adaptive refinement   = 0`

The log.txt and a prm that be can quickly run on one processor are attached. Let me know how I can help with additional testing!

[test_weighted_particle_average_refinement.prm.txt](https://github.com/user-attachments/files/15874642/test_weighted_particle_average_refinement.prm.txt)
[log.txt](https://github.com/user-attachments/files/15874644/log.txt)

gassmoeller commented:
@naliboff: Sorry about that I clearly didnt do enough testing for this interpolator, adaptive refinement was broken. I think I fixed it now and added a test. Could you give your model another try? I encountered another problem with your model that might be a bug in deal.II. If you crash still try changing the number of particles per direction to 6 (from 7). I will look into that more.

naliboff commented:
@gassmoeller - Thanks for the fix and the model runs successfully now with both adaptive refinement and either 6 or 7 initial particles per direction. Is there any other additional testing I can do in the short term? I'll look more carefully at the effect of the weighted average scheme on convergence on the complex models in the coming weeks.

tjhei reviewed (approved):
Looks good to me. Anything else you want to do here, @gassmoeller ?

gassmoeller commented:
This is ready from my side. Just a quick note that at present this interpolator is quite slow (about twice as slow than the linear least squares interpolator, which does a much more complicated operation). Presumably this is because it has to create the `std::set` of neighbor cells for every cell. I have some ideas for how to speed this up, but it is not a terrible cost at the moment and the functionality should be fine as is, so I think we can merge this PR and address the performance in a follow up.
